dali/operators/reader/loader/video/frames_decoder.cc

```diff
@@ -0,0 +1,259 @@
+        return;
+    }
+  }
+
+  DALI_FAIL(make_string("Could not find a valid video stream in a file ", filename_));
+}
+
+FramesDecoder::FramesDecoder(const std::string &filename)
+    : av_state_(std::make_unique<AvState>()), filename_(filename) {
+  av_state_->ctx_ = avformat_alloc_context();
+  DALI_ENFORCE(av_state_->ctx_, "Could not alloc avformat context");
+
+  int ret = avformat_open_input(&av_state_->ctx_, filename.c_str(), nullptr, nullptr);
+  DALI_ENFORCE(ret == 0, make_string("Failed to open video file at path ", filename, "due to ",
+                                     detail::av_error_string(ret)));
+
+  FindVideoStream();
+  InitAvState();
+  BuildIndex();
+}
+
+void FramesDecoder::BuildIndex() {
+  // TODO(awolant): Optimize this function for: 
+  //  - CFR
+  //  - index present in the header
+
+  int last_keyframe = -1;
+  while (ReadRegularFrame(nullptr, false)) {
+    IndexEntry entry;
+    entry.is_keyframe = av_state_->frame_->key_frame;
+    entry.pts = av_state_->frame_->pts;
+    entry.is_flush_frame = false;
+    
+    if (entry.is_keyframe) {
+      last_keyframe = index_.size();
+    }
+    entry.last_keyframe_id = last_keyframe;
+    index_.push_back(entry);
+  }
+  while (ReadFlushFrame(nullptr, false)) {
+    IndexEntry entry;
+    entry.is_keyframe = false;
+    entry.pts = av_state_->frame_->pts;
+    entry.is_flush_frame = true;
+    entry.last_keyframe_id = last_keyframe;
+    index_.push_back(entry);
+  }
+  Reset();
+}
+
+void FramesDecoder::CopyToOutput(uint8_t *data) {
+  uint8_t *dest_[4] = {data, nullptr, nullptr, nullptr};
+  int dest_linesize_[4] = {av_state_->frame_->width * Channels(), 0, 0, 0};
+
+  int ret = sws_scale(
+    av_state_->sws_ctx_,
+    av_state_->frame_->data,
+    av_state_->frame_->linesize,
+    0,
+    av_state_->frame_->height,
+    dest_,
+    dest_linesize_);
+
+  DALI_ENFORCE(
+    ret >= 0,
+    make_string("Could not convert frame data to RGB: ", detail::av_error_string(ret)));
+}
+
+void FramesDecoder::LazyInitSwContext() {
+  if (av_state_->sws_ctx_ == nullptr) {
+    av_state_->sws_ctx_ = sws_getContext(
+      Width(), 
+      Height(), 
+      av_state_->codec_ctx_->pix_fmt, 
+      Width(), 
+      Height(), 
+      AV_PIX_FMT_RGB24, 
+      SWS_BILINEAR, 
+      nullptr, 
+      nullptr, 
+      nullptr);
+    DALI_ENFORCE(av_state_->sws_ctx_, "Could not create sw context");
+  }
+}
+
+bool FramesDecoder::ReadRegularFrame(uint8_t *data, bool copy_to_output) {
+  int ret = -1;
+  while (av_read_frame(av_state_->ctx_, av_state_->packet_) >= 0) {
+    if (av_state_->packet_->stream_index != av_state_->stream_id_) {
+      continue;
+    }
+
+    ret = avcodec_send_packet(av_state_->codec_ctx_, av_state_->packet_);
+    DALI_ENFORCE(
+      ret >= 0,
+      make_string("Failed to send packet to decoder: ", detail::av_error_string(ret)));
+
+    ret = avcodec_receive_frame(av_state_->codec_ctx_, av_state_->frame_);
+
+    if (ret == AVERROR(EAGAIN)) {
```

Comment:
Is it possible to get fatal error and throw here?

Reply:
From what I understand, this error basically tells you that decoder doesn't have enough packets and you need to feed another one and try again.

Reply:
I'm not asking about this particular value, but we test only `EAGAIN` and `AVERROR_EOF` but not any other. I think if there is any other value we should throw.